### DISCUSSION TITLE
Out of Band Commissioning

### DISCUSSION THREAD

**Discussion by ArminBreiteneder-eaton** [openthread/openthread#11016] [Q&A]
Hello!

What is the best approach for doing out-of-band commissiong?
Currently I am supplying the Networkkey, Networkname, PanID, Extended PanID and Channel to the device using BLE. I am putting these infos to a otOperationalDataset structure and feeding it into the function otDatasetSetActive(). Then I call otIp6SetEnabled() and otThreadSetEnabled().

Using this method it mostly takes ~15 seconds until the new device joined the network. In some rare cases it is ~5 seconds. They mostly start to join the network with Link Request which are not answered by the existing network. After these unsuccessful tries it uses Parent request to finally join the network correctly.

Is there some optimization possible?

**Comment by abtink** [comment]
Some suggestions:
- If you can, providing the entire Active Operation Dataset can be better (can be read as a sequence of TLV (data blob) and can set it as such on the new device). Please check [`otDatasetGetActiveTlvs()`](https://github.com/openthread/openthread/blob/7b72e9778253a87c404cc27af85be5cf12b0ef11/include/openthread/dataset.h#L338-L347) and [`otDatasetSetActiveTlvs()`](https://github.com/openthread/openthread/blob/7b72e9778253a87c404cc27af85be5cf12b0ef11/include/openthread/dataset.h#L377-L399).
- May help if you factory-reset before setting the new dataset. This may help speed up the initial attach due to device not trying to restore its previous role/RLOC16 ([`otInstanceFactoryReset`](https://github.com/openthread/openthread/blob/7b72e9778253a87c404cc27af85be5cf12b0ef11/include/openthread/instance.h#L273-L278) and/or [`otInstanceErasePersistentInfo`](https://github.com/openthread/openthread/blob/7b72e9778253a87c404cc27af85be5cf12b0ef11/include/openthread/instance.h#L291-L300))

**Reply by ArminBreiteneder-eaton** [reply]
After some test we realized that it is much faster if a device is in factory defaults without active network. I guess there is no option that handles the reset of the current network in the OpenThread library without a restart of the device?

**Reply by jwhui** [reply]
After clearing the non-volatile settings, you can call `otInstanceFinalize()` then `otInstanceInit()` or `otInstanceInitSingle()` to reset the OpenThread instance.
